iwyu.cc

```diff
@@ -4272,22 +4296,41 @@ class IwyuAstConsumer
     }
 
     instantiated_template_visitor_.ScanInstantiatedFunction(
-        callee, parent_type,
-        current_ast_node(), resugar_map);
+        callee, parent_type, current_ast_node(), resugar_map,
+        ExtractProvidedComponents(resugar_map));
     return true;
   }
 
   // --- Handler declared in IwyuBaseASTVisitor.
 
-  void ReportTplSpecComponentTypes(const TemplateSpecializationType* type) {
+  void ReportTplSpecComponentTypes(const TemplateSpecializationType* type,
+                                   const set<const Type*>& blocked_types) {
     const map<const Type*, const Type*> resugar_map =
         GetTplTypeResugarMapForClass(type);
     ASTNode node(type);
     node.SetParent(current_ast_node());
-    instantiated_template_visitor_.ScanInstantiatedType(&node, resugar_map);
+    set<const Type*> merged_blocked = ExtractProvidedComponents(resugar_map);
+    merged_blocked.insert(blocked_types.begin(), blocked_types.end());
+    instantiated_template_visitor_.ScanInstantiatedType(&node, resugar_map,
+                                                        merged_blocked);
   }
 
  private:
+  set<const Type*> ExtractProvidedComponents(
+      const map<const Type*, const Type*>& resugar_map) const {
+    set<const Type*> result;
+    for (const auto [_, sugar_type] : resugar_map) {
```

Comment:
Very cool that destructuring assignments are becoming mainstream :).

I searched through the llvm codebase, though, and only found a single use besides tests. So I think we should stick with the old-school `for (const auto& item : resugar_map)` for now, so as not to alienate us old C++ folks.

Reply:
Maybe, it's because LLVM switched to C++17 relatively recently.